src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,712 @@
+                        values[i] = array[i].Value;
+                    }
+
+                    return values;
+                }
+
+                return _dictionaryStorage!.Values;
+            }
+        }
+
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+
+        /// <inheritdoc />
+        void ICollection<KeyValuePair<TKey, TValue>>.Add(KeyValuePair<TKey, TValue> item)
+        {
+            if (_arrayStorage != null)
+            {
+                Add(item.Key, item.Value);
+                return;
+            }
+
+            ((ICollection<KeyValuePair<TKey, TValue>>)_dictionaryStorage!).Add(item);
+            return;
+        }
+
+        /// <inheritdoc />
+        public void Add(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_arrayStorage != null)
+            {
+                EnsureCapacity(_count + 1);
+
+                if (_dictionaryStorage != null)
+                {
+                    Debug.Assert(_arrayStorage == null);
+                    _dictionaryStorage.Add(key, value);
+                    return;
+                }
+
+                if (ContainsKeyArray(key))
+                {
+                    throw new ArgumentException($"An element with the key '{key}' already exists in the {nameof(AdaptiveCapacityDictionary<TKey, TValue>)}.", nameof(key));
+                }
+
+                _arrayStorage[_count] = new KeyValuePair<TKey, TValue>(key, value);
+                _count++;
+                return;
+            }
+
+            _dictionaryStorage!.Add(key, value);
+            return;
+        }
+
+        /// <inheritdoc />
+        public void Clear()
+        {
+            if (_dictionaryStorage != null)
+            {
+                _dictionaryStorage.Clear();
+            }
+
+            if (_count == 0)
+            {
+                return;
+            }
+            if (_arrayStorage != null)
+            {
+                Array.Clear(_arrayStorage, 0, _count);
+                _count = 0;
+            }
+        }
+
+        /// <inheritdoc />
+        bool ICollection<KeyValuePair<TKey, TValue>>.Contains(KeyValuePair<TKey, TValue> item)
+        {
+            if (_dictionaryStorage != null)
+            {
+                return ((ICollection<KeyValuePair<TKey, TValue>>)_dictionaryStorage).Contains(item);
+            }
+
+            return TryGetValue(item.Key, out var value) && EqualityComparer<object>.Default.Equals(value, item.Value);
+        }
+
+        /// <inheritdoc />
+        public bool ContainsKey(TKey key)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_dictionaryStorage != null)
+            {
+                return _dictionaryStorage.ContainsKey(key);
+            }
+
+            return ContainsKeyCore(key);
```

Comment:
Here I'd be more aggressive and prioritize the small sizes.

```suggestion
            if (_dictionaryStorage is null)
            {
                return ContainsKeyArray(key);
            }

            return _dictionaryStorage.ContainsKey(key);
```
For codegen the small sizes are in the hot-path, whilst the fallback is moved down to a colder region.
Additional the cpu's branch predictor prefers the fall-through path when there's not history.

In the micro-benchmarks this effect won't show up, as the branch predictor has history and the I-caches are hot. The benefit should be in real-world usage -- but honestly, it's mini.

The method itself will inline, so aggressive-inline isn't necessary, but you can add if you want. I don't know...